docs/src/dev/understanding_traces.md

```diff
@@ -0,0 +1,39 @@
+# Understanding Traces
+
+yk can print the traces it has created to `stderr` to help with debugging.
+However, these traces are often lengthy, and not always easy to understand.
+This section briefly explains how to get yk to print its traces, and how
+to make them a bit easier to understand.
+
+
+## Producing a trace
+
+The `YKD_PRINT_IR` environment variable determines whether yk prints traces to
+`stderr` or not. If `jit-pre-opt` is specified, the traces will be printed before
+optimisation; if `jit-post-opt` is specified, the traces will be printed after
+optimisation. `jit-pre-opt` and `jit-post-opt` can give you different insights,
+so it is often worth checking both.
+
+
+## trace_chewer
+
+`trace_chewer` is a small program included with yk which can help you
+understand a trace.
+
+
+### simplify
+
+`trace_chewer simplify <file-name>` takes as input a trace and converts
+it into a straight-line trace. It removes LLVM declarations and guard
+failure blocks, neither of which are enlightening. It converts LLVM
+branches into simpler `guard_true(%var)` or `guard_false(%var)` statements
+(i.e. `guard_true(%3)` means "if the boolean value stored in `%3` is true,
+then continue executing the trace, otherwise deoptimise back to the
+interpreter).
+
+If you specify `-` as the filename, `trace_chewer` will read from stdin. This
+means that you can simplify traces without saving them to disk:
+
+```
+YKD_PRINT_IR=jit-post-opt lua f.lua | trace_chewer simplify -
```

Comment:
So this needs a 2>&1? But what would it do with program output interleaved/prepended/postpended with the module?

I guess pre/post is fine as you are looking for the print delimeters. Is interleaving possible? Not sure.

Reply:
Already fixed in https://github.com/ykjit/yk/pull/654/commits/b7b4d60a61d4d52cdf3bc3a4f8e738e1797ab9a0.